### ISSUE TITLE
Possible to make one-line @Data declarations?

### ISSUE THREAD

**Issue by thorning** [HerringtonDarkholme/av-ts#11]
Would it be possible to make the `@Data` decorator work on individual properties?

```
class TestComponent extends Vue {
  normal = 'normal';
  @Data test = () => new Test();
  @Data own = function () {return this.propValue} //would need this syntax to correctly bind this
  @Data shared = () => store.sharedObject
}
```

I have not worked with decorators before, so don't know how much is possible. I guess the biggest problem is the type signatures, as `this.test` needs to be of type `Test` not `() => Test`.

Current workaround is to set all data in the data function, and simply declare the variables on the class for typings:

```
class TestComponent extends Vue {
  normal:string;
  test:Test;
  own:PropType;
  shared:SharedObject;
  @Data data() {
    return {
      normal : 'normal',
      test : new Test(),
      own : this.propValue,
      shared : store.sharedObject,
    }
  }
}
```

**Comment by HerringtonDarkholme** [comment]
Thanks for using av-ts, again.

I'm sorry, decorator cannot change property type. It's TypeScript's feature.

I'm considering the pain point of declaring multiple heterogeneous data. Do you have any idea on improving this?

**Comment by HerringtonDarkholme** [comment]
Let me summarize something:

1. literal value: already supported well
2. class/function object: lost class information. This probably makes user need to declare interface while not necessary. Also runtime check is impossible.
3. prop value: supported only in Data function. Almost impossible in property style.
4. shared object: impossible to support in property declaration style.

**Comment by thorning** [comment]
It is me who is thankful you made and maintain this :)

My first thought would be to do something like:
`@Data get own() {return this.propValue}`

where the getter would be used when you construct the data function, but the data function sets it as a normal (non-getter) property.

This would keep the type signature, but as a readonly, which doesn't really help.


Tbh. just using the `@Data` function when needed is not that big a deal unless you have a large amount of properties, which goes against the component architecture anyway.

**Comment by ktsn** [comment]
Hi, I actually implemented this feature on `vue-class-component`.
I hope this can help you.
https://github.com/vuejs/vue-class-component/pull/29

**Comment by HerringtonDarkholme** [comment]
@ktsn Thanks for giving av-ts a look!
It is very clever to use enumerable to exclude props! 
This library has additional concern that `@prop = p(Number)` will have to be re-evluated in [constructor call.](https://github.com/vuejs/vue-class-component/pull/29/files#diff-3b68c45ddf0810c826b716800f53b92cR27) But I think I can set a flag to bail out execution...

Thanks! I'll try to implement this soon!

**Comment by HerringtonDarkholme** [comment]
I have pushed a new implementation of this. https://github.com/HerringtonDarkholme/av-ts/commit/ed0796e30614206778fd9c6ac9b24ba69a9e2370

@thorning @ktsn it would be my honor if you can take a look on it.

**Comment by ktsn** [comment]
@HerringtonDarkholme looks good to me 🎉

**Comment by HerringtonDarkholme** [comment]
published as 0.5.0  🎉

**Comment by thorning** [comment]
This looks good to me. So far it have behaved as I intuitively thought it would, including constructors, factory methods and shared stores.
